beacon.go

```diff
@@ -17,110 +17,164 @@
-		return
+func validateServerConfig() error {
+	if config.projectID == "" {
+		return fmt.Errorf("%s must be specified", projectKey)
+	}
+	if config.tableId == "" {
+		return fmt.Errorf("%s must be specified", bqTableKey)
 	}
+	return nil
+}
 
-	client, err := google.DefaultClient(c, genomics.GenomicsReadonlyScope)
-	if err != nil {
-		http.Error(w, "Invalid server configuration", http.StatusInternalServerError)
+func parseInput(r *http.Request) (string, string, int64, error) {
+	refName := r.FormValue("chromosome")
+	if refName == "" {
+		return "", "", 0, newBadRequestError("parsing chromosome name", errors.New("value is required"))
+	}
+	allele := r.FormValue("allele")
+	if refName == "" {
+		return "", "", 0, newBadRequestError("parsing allele name", errors.New("value is required"))
 	}
-	genomicsService, err := genomics.New(client)
+	coord, err := strconv.ParseInt(r.FormValue("coordinate"), 10, 64)
 	if err != nil {
-		http.Error(w, "Invalid server configuration", http.StatusInternalServerError)
+		return "", "", 0, newBadRequestError("parsing coordinate", err)
 	}
-	variantsService := genomics.NewVariantsService(genomicsService)
+	return refName, allele, coord, nil
+}
 
+func writeResponse(w http.ResponseWriter, exists bool) error {
 	type beaconResponse struct {
 		XMLName struct{} `xml:"BEACONResponse"`
 		Exists  bool     `xml:"exists"`
 	}
 	var resp beaconResponse
-
-	for {
-		searchResponse, err := variantsService.Search(search).Do()
-		if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-			return
-		}
-
-		for _, variant := range searchResponse.Variants {
-			if search.Start != variant.Start {
-				continue
-			}
-			if allele == variant.ReferenceBases {
-				resp.Exists = true
-			} else {
-				for _, base := range variant.AlternateBases {
-					if base == allele {
-						resp.Exists = true
-						break
-					}
-				}
-			}
-		}
-
-		if resp.Exists || searchResponse.NextPageToken == "" {
-			break
-		}
-		search.PageToken = searchResponse.NextPageToken
-	}
+	resp.Exists = exists
 
 	w.Header().Set("Content-Type", "application/xml")
 	enc := xml.NewEncoder(w)
 	enc.Indent("", "  ")
-	if err = enc.Encode(resp); err != nil {
-		http.Error(w, "Failed writing response", http.StatusInternalServerError)
+	if err := enc.Encode(resp); err != nil {
+		return fmt.Errorf("serializing response: %v", err)
+	}
+	return nil
+}
+
+type apiError struct {
+	code  int
+	cause error
+}
+
+func (err *apiError) Error() string {
+	return fmt.Sprintf("%d: %v", err.code, err.cause)
+}
+
+func newHttpError(code int, context string, err error) error {
+	return &apiError{code, fmt.Errorf("%s: %v", context, err)}
+}
+
+func newBadRequestError(context string, err error) error {
+	return newHttpError(http.StatusBadRequest, context, fmt.Errorf("invalid input: %v", err))
+}
+
+// writeError writes a bare HTTP error describing err to w.
+func writeError(w http.ResponseWriter, err error) {
+	if err, ok := err.(*apiError); ok {
```

Comment:
code := http.StatusInternalServerError
if err, ok := err.(*apiError); ok {
  code = err.code
}
http.Error(w, fmt.Sprintf("%s: %v", http.StatusText(code), err), code)

Then you can drop `writeHTTPError`

Reply:
not applicable since I reverted the errors to what they used to be